File: src/main/java/com/fengsheng/card/DiaoBao.java
package com.fengsheng.card;

import com.fengsheng.*;
import com.fengsheng.phase.FightPhaseIdle;
import com.fengsheng.phase.OnUseCard;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

public class DiaoBao extends Card {
    private static final Logger log = Logger.getLogger(DiaoBao.class);

    public DiaoBao(int id, Common.color[] colors, Common.direction direction, boolean lockable) {
        super(id, colors, direction, lockable);
    }

    public DiaoBao(int id, Card card) {
        super(id, card);
    }

    /**
     * 仅用于“作为调包使用”
     */
    DiaoBao(Card originCard) {
        super(originCard);
    }

    @Override
    public Common.card_type getType() {
        return Common.card_type.Diao_Bao;
    }

    @Override
    public boolean canUse(Game g, Player r, Object... args) {
        if (r == g.getJinBiPlayer()) {
            log.error("你被禁闭了，不能出牌");
            return false;
        }
        if (g.getQiangLingTypes().contains(getType())) {
            log.error("调包被禁止使用了");
            return false;
        }
        if (!(g.getFsm() instanceof FightPhaseIdle fsm) || r != fsm.whoseFightTurn) {
            log.error("调包的使用时机不对");
            return false;
        }
        return true;
    }

    @Override
    public void execute(final Game g, final Player r, Object... args) {
        final var fsm = (FightPhaseIdle) g.getFsm();
        log.info(r + "使用了" + this);
        r.deleteCard(this.id);
        Fsm resolveFunc = () -> {
            Card oldCard = fsm.messageCard;
            g.getDeck().discard(oldCard);
            fsm.messageCard = this.getOriginCard();
            fsm.isMessageCardFaceUp = false;
            fsm.whoseFightTurn = fsm.inFrontOfWhom;
            for (Player player : g.getPlayers()) {
                if (player instanceof HumanPlayer p) {
                    var builder = Fengsheng.use_diao_bao_toc.newBuilder();
                    builder.setOldMessageCard(oldCard.toPbCard()).setPlayerId(p.getAlternativeLocation(r.location()));
                    if (p == r) builder.setCardId(this.id);
                    p.send(builder.build());
                }
            }
            return new ResolveResult(fsm, true);
        };
        g.resolve(new OnUseCard(fsm.whoseTurn, r, null, this, Common.card_type.Diao_Bao, r, resolveFunc));
    }

    @Override
    public String toString() {
        return Card.cardColorToString(colors) + "调包";
    }

    public static boolean ai(FightPhaseIdle e, Card card) {
        Player player = e.whoseFightTurn;
        var colors = e.messageCard.getColors();
        if (e.inFrontOfWhom == player && (e.isMessageCardFaceUp || player == e.whoseTurn) && colors.size() == 1 && colors.get(0) != Common.color.Black)
            return false;
        if (ThreadLocalRandom.current().nextInt(4) != 0)
            return false;
        GameExecutor.post(player.getGame(), () -> card.execute(player.getGame(), player), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/card/DiaoBao.kt
package com.fengsheng.card

import com.fengsheng.Game
import com.fengsheng.HumanPlayer
import com.fengsheng.Player
import com.fengsheng.phase.FightPhaseIdle
import com.fengsheng.phase.OnFinishResolveCard
import com.fengsheng.phase.ResolveCard
import com.fengsheng.protos.Common.card_type.Diao_Bao
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.direction
import com.fengsheng.protos.Common.phase.Fight_Phase
import com.fengsheng.protos.notifyPhaseToc
import com.fengsheng.protos.useDiaoBaoToc
import com.fengsheng.skill.cannotPlayCard
import org.apache.logging.log4j.kotlin.logger

class DiaoBao : Card {
    constructor(id: Int, colors: List<color>, direction: direction, lockable: Boolean) :
            super(id, colors, direction, lockable)

    constructor(id: Int, card: Card) : super(id, card)

    /**
     * 仅用于“作为调包使用”
     */
    internal constructor(originCard: Card) : super(originCard)

    override val type = Diao_Bao

    override fun canUse(g: Game, r: Player, vararg args: Any): Boolean {
        if (r.cannotPlayCard(type)) {
            logger.error("你被禁止使用调包")
            r.sendErrorMessage("你被禁止使用调包")
            return false
        }
        if (r !== (g.fsm as? FightPhaseIdle)?.whoseFightTurn) {
            logger.error("调包的使用时机不对")
            r.sendErrorMessage("调包的使用时机不对")
            return false
        }
        return true
    }

    override fun execute(g: Game, r: Player, vararg args: Any) {
        val fsm = g.fsm as FightPhaseIdle
        logger.info("${r}使用了$this")
        r.deleteCard(id)
        val resolveFunc = { _: Boolean ->
            val oldCard = fsm.messageCard
            g.deck.discard(oldCard)
            for (player in g.players) {
                if (player is HumanPlayer) {
                    player.send(useDiaoBaoToc {
                        oldMessageCard = oldCard.toPbCard()
                        playerId = player.getAlternativeLocation(r.location)
                        if (player === r) cardId = id
                    })
                }
            }
            val newFsm = fsm.copy(
                messageCard = getOriginCard(),
                isMessageCardFaceUp = false,
                whoseFightTurn = fsm.inFrontOfWhom
            )
            for (p in g.players) { // 解决客户端动画问题
                if (p is HumanPlayer) {
                    p.send(notifyPhaseToc {
                        currentPlayerId = p.getAlternativeLocation(newFsm.whoseTurn.location)
                        messagePlayerId = p.getAlternativeLocation(newFsm.inFrontOfWhom.location)
                        waitingPlayerId = p.getAlternativeLocation(newFsm.whoseFightTurn.location)
                        currentPhase = Fight_Phase
                    })
                }
            }
            OnFinishResolveCard(
                fsm.whoseTurn,
                r,
                null,
                getOriginCard(),
                Diao_Bao,
                newFsm,
                discardAfterResolve = false
            )
        }
        g.resolve(ResolveCard(fsm.whoseTurn, r, null, getOriginCard(), Diao_Bao, resolveFunc, fsm))
    }

    override fun toString(): String {
        return "${cardColorToString(colors)}调包"
    }
}